Store a version in the database and use it for database activity on startup

mikehardy opened the pull request:
## Pull Request template

## Purpose / Description
The application has a version in the preferences right now, and uses that to decide if special activity (migrations, integrity checks etc) need to happen on startup.

It's possible that preferences are backed up and restored from Google though, leaving the current application in an inconsistent state where the database (which is not backed up with google services) may have had different things done to it than the preferences.

This adds an ankidroid version to the Collection configuration, and uses it to decide what to do on startup.

For the special case where we have no version in the database (the first time this code runs, or if there is a new collection), we assume that we need to run a database check and then set the version to current.

## Fixes
#5106

## Approach
This builds on #5171 which was just refactoring/setup to try to avoid messing things up with this one. I kept them separate even though they are connected, so each commit / file history would be coherent historically, and ideally is easier to review? If #5171 is merged before this I'll either rebase this one, or it will rebase itself on merge

The approach is as discussed in the relevant issue as well as the first attempt #5120 which was insufficient - store a version in deck conf

## How Has This Been Tested?

Tested in emulators as well as with automated robolectric coverage in unit tests (`./gradlew jacocoUnitTestReport` then open the report HTML file in the build/reports directory)

## Learning (optional, can help others)
Nothing special with this one

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -875,153 +875,177 @@ public void addNote() {
 
-    private void showStartupScreensAndDialogs(SharedPreferences preferences, int skip) {
+    @VisibleForTesting
+    protected void showStartupScreensAndDialogs(SharedPreferences preferences, int skip) {
```
Shorter than it was before at least...

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:
```diff
@@ -875,153 +875,177 @@ public void addNote() {
 
-    private void showStartupScreensAndDialogs(SharedPreferences preferences, int skip) {
+    @VisibleForTesting
+    protected void showStartupScreensAndDialogs(SharedPreferences preferences, int skip) {
```
Less complex than it was before at least...

mikehardy commented:
The database integrity check progress touched the same area so caused a small conflict. I just rebased and repushed. @timrae not meaning to pester - just noting that codacy will see two problems with this but I am not going to resolve them, it's ready for review as is. The problems were "really long method" and "npath complexity" - reducing both issues was an aim for the PR and I chopped them in about half but they were so gnarly before that going all the way this time was too much. So they get flagged...

mikehardy commented:
@timrae this continues to be clean vs master on a rebase / force-push, and the [testing support](https://github.com/ankidroid/Anki-Android/pull/5172/files#diff-6758ada5e8ea62bf68c113fe089694e5R97) continues to give me confidence it’s ready to go. Just needs review.

timrae commented:
I'm not really a fan of the unit tests accessing private state variables, as it will cause issues anytime we try to change anything, but for now we might as well have them in there. Don't be shocked if I remove them rather than fix them when the time comes though ;)

mikehardy commented:
I'm always open to suggestions on how to improve testing. I have definitely found that on my projects with tests vs without, more tests seems to lead to more collaborators and more successful PRs, worth pondering for at least a moment before nuking :-). Thanks for the merge, hopefully it will have more success than my previous foray into the show startup dialogs snarl

timrae commented:
I was a bit weirded out on installation of the latest alpha to get an *error report being sent to developers....* or whatever snackbar after the integrity check finished. I'm not sure whether or not we should be sending that information to the server after the db check. Probably it's not necessary, and some people don't like it.

```
Stack Trace | java.lang.Exception: Problem list (limited to first 10)at com.ichi2.libanki.Collection.logProblems(Collection.java:5)at com.ichi2.libanki.Collection.fixIntegrity(Collection.java:116)at com.ichi2.async.DeckTask.doInBackgroundCheckDatabase(DeckTask.java:5)at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:45)at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:1)at android.os.AsyncTask$2.call(AsyncTask.java:333)at java.util.concurrent.FutureTask.run(FutureTask.java:266)at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)at java.lang.Thread.run(Thread.java:764)
-- | --
{additionalInfo:Reviews had incorrect due date., origin:Collection.fixIntegrity()}
```

mikehardy commented:
Oh yes - similar to analytics that was so I could see if there were any other latent corruption patterns similar to the ordinals from card template editor. I have enough of them from the alpha users that I feel like I can say there does not appear to be anything special, you could delete this or I could PR to delete this. On the other hand, this does mean that your collection was corrupt?

mikehardy commented:
¨Reviews had incorrect due date.¨ actually that shows up a few times

timrae commented:
PR would be appreciated!

mikehardy commented:
I´ll canvas for a bit more and log findings here then if it is removed there is at least a record

Fixed 6 card(s) with invalid properties.
Deleted 1 note(s) with missing no cards.
Deleted 1 note(s) with missing no cards.
Fixed 7 cards with v2 scheduler bug. Fixed 8 review history entries with v2 scheduler bug.
Deleted 3 note(s) with missing no cards.
Deleted 5494 note(s) with missing no cards.
Deleted 25387 note(s) with missing no cards.
Fixed 197 review history entries with v2 scheduler bug.
Reviews had incorrect due date. 
Fixed 1 cards with v2 scheduler bug. 
Fixed 15 review history entries with v2 scheduler bug.
Deleted 197 card(s) with missing template.
Deleted 2 note(s) with missing no cards.
Deleted 2 note(s) with missing no cards.
Deleted 1 note(s) with missing no cards.
Reviews had incorrect due date.

Okay, those are the ones I picked out of the most recent 14,000 ACRA reports, going back about 2 months.

I do see a pattern ;-) - there is something going on with review dates and ability to create notes without cards or delete cards orphaning notes.

I´ll PR something to disable the check since it´s easy but leave it to you to think about whether the info is valuable (I think it might be?) thus whether to merge or reject

Separately, may be useful to create an issue tracking note orphaning and/or tracking review date correction - where an audit between what the DB is doing and who else touches that data in a maybe-corrupting way is noted, and tagged help wanted for someone that wanted to learn the codebase but in a motivated way ? A couple of those found traction recently
